lib/MusicBrainz/Server/Controller/Account/Subscriptions/Collection.pm

```diff
@@ -21,7 +21,11 @@ before add => sub
         $c->detach('/error_400');
     }
 
-    $c->detach('/error_403') if (!$entity->public && $c->user->id != $entity->editor_id);
+    my $is_collection_owner = $c->user->id == $entity->editor_id;
+    my $is_collection_collaborator = $c->user_exists &&
+        $c->model('Collection')->is_collection_collaborator($c->user->id, $entity->id);
```

Comment:
The `add` method from the role has `RequireAuth`, so I think we don't need to check `$c->user_exists` here (and we already reference `$c->user->id` above without checking it).

It would be nice to avoid running this query altogether if we already know either of `$entity->public` or `$is_collection_owner` are true. Perhaps something like:

```Perl
my $is_authorized = (
    $entity->public ||
    $c->user->id == $entity->editor_id ||
    $c->model('Collection')->is_collection_collaborator($c->user->id, $entity->id)
);
```